pkg/queue/cluster_queue_impl.go

```diff
@@ -55,39 +72,99 @@ func (c *ClusterQueueImpl) Cohort() string {
 func (c *ClusterQueueImpl) AddFromQueue(q *Queue) bool {
 	added := false
 	for _, info := range q.items {
-		if c.pushIfNotPresent(info) {
+		if c.heap.PushIfNotPresent(info) {
 			added = true
 		}
 	}
 	return added
 }
 
+func (c *ClusterQueueImpl) PushOrUpdate(wInfo *workload.Info) {
+	key := workload.Key(wInfo.Obj)
+	oldInfo := c.inadmissibleWorkloads[key]
+	if oldInfo != nil {
+		// update in place if the workload was inadmissible and didn't change
+		// to potentially become admissible.
+		if equality.Semantic.DeepEqual(oldInfo.Obj.Spec, wInfo.Obj.Spec) {
+			c.inadmissibleWorkloads[key] = wInfo
+			return
+		}
+		// otherwise move or update in place in the queue.
+		delete(c.inadmissibleWorkloads, key)
+	}
```

Comment:
We can, I wanted to do that initially (and already made the changes to the workload controller to update the workload status), but left it out of this PR to simplify things since this PR is already long.